sdlib/d/sync/sharedlock.d

```diff
@@ -0,0 +1,340 @@
+		assert(count == Exclusive);
+		count = 0;
+	}
+}
+
+unittest sharedLocks {
+	static runThread(void* delegate() dg) {
+		static struct Delegate {
+			void* ctx;
+			void* function(void*) fun;
+		}
+
+		auto x = *(cast(Delegate*) &dg);
+
+		import core.stdc.pthread;
+		pthread_t tid;
+		auto r = pthread_create(&tid, null, x.fun, x.ctx);
+		assert(r == 0, "Failed to create thread!");
+
+		return tid;
+	}
+
+	shared SharedLock lock;
+	assert(lock.count == 0);
+	lock.sharedLock();
+	assert(lock.count == 1);
+	assert(!lock.mutex.isHeld());
+	lock.sharedLock();
+	assert(lock.count == 2);
+	lock.sharedUnlock();
+	assert(lock.count == 1);
+
+	import d.sync.atomic;
+	shared Mutex mutex;
+	shared Atomic!uint numLocked;
+	bool exitThread;
+	void* run() {
+		lock.sharedLock();
+		scope(exit) lock.sharedUnlock();
+		numLocked.fetchAdd(1);
+		mutex.lock();
+		scope(exit) mutex.unlock();
+		bool shouldExit() {
+			return exitThread;
+		}
+
+		mutex.waitFor(shouldExit);
+		return null;
+	}
+
+	import core.stdc.pthread;
+	pthread_t[50] tids;
+	foreach (i; 0 .. tids.length) {
+		tids[i] = runThread(run);
+	}
+
+	{
+		bool allThreadsReady() {
+			return numLocked.load() == tids.length;
+		}
+
+		mutex.lock();
+		scope(exit) mutex.unlock();
+		mutex.waitFor(allThreadsReady);
+	}
+
+	assert(lock.count == 51);
+	{
+		mutex.lock();
+		scope(exit) mutex.unlock();
+		exitThread = true;
+	}
+
+	void* ret;
+	foreach (i; 0 .. tids.length) {
+		pthread_join(tids[i], &ret);
+	}
+
+	assert(lock.count == 1);
+	lock.sharedUnlock();
+	assert(lock.count == 0);
+}
+
+unittest exclusiveLock {
+	static runThread(void* delegate() dg) {
+		static struct Delegate {
+			void* ctx;
+			void* function(void*) fun;
+		}
+
+		auto x = *(cast(Delegate*) &dg);
+
+		import core.stdc.pthread;
+		pthread_t tid;
+		auto r = pthread_create(&tid, null, x.fun, x.ctx);
+		assert(r == 0, "Failed to create thread!");
+
+		return tid;
+	}
+
+	shared SharedLock lock;
+	lock.exclusiveLock();
+	assert(lock.count == SharedLock.Exclusive);
+	lock.exclusiveUnlock();
+	assert(lock.count == 0);
+	lock.sharedLock();
+	assert(lock.count == 1);
+
+	// Attempt to exclusive lock while the shared lock is held.
+	import core.stdc.pthread;
+	import d.sync.atomic;
+	shared Atomic!uint state;
+	void* takeExclusiveLock() {
+		state.fetchAdd(1);
+		lock.exclusiveLock();
+		state.fetchAdd(1);
+		return cast(void*) cast(size_t) lock.count;
+	}
+
+	auto writer = runThread(takeExclusiveLock);
+	{
+		bool triedExclusiveLock() {
+			return lock.count >= SharedLock.Exclusive;
+		}
+
+		lock.mutex.lock();
+		scope(exit) lock.mutex.unlock();
+
+		lock.mutex.waitFor(triedExclusiveLock);
+		assert(lock.count == SharedLock.Exclusive + 1);
+		assert(state.load() == 1);
+	}
+
+	lock.sharedUnlock();
+	assert(lock.count == SharedLock.Exclusive);
+	void* ret;
+	pthread_join(writer, &ret);
+	assert(cast(size_t) ret == SharedLock.Exclusive);
+	assert(state.load() == 2);
+	lock.exclusiveUnlock();
+	assert(lock.count == 0);
```

Comment:
Give me some new lines, don't mix up thread operation with checking the locks and so on.